collaborative-plonk/src/round4.rs

```diff
@@ -0,0 +1,240 @@
+    {
+        Self {
+            beta: round3_challenges.beta,
+            gamma: round3_challenges.gamma,
+            alpha: round3_challenges.alpha,
+            xi,
+        }
+    }
+}
+
+pub(super) struct Round4Proof<P: Pairing> {
+    pub(super) commit_a: P::G1,
+    pub(super) commit_b: P::G1,
+    pub(super) commit_c: P::G1,
+    pub(super) commit_z: P::G1,
+    pub(super) commit_t1: P::G1,
+    pub(super) commit_t2: P::G1,
+    pub(super) commit_t3: P::G1,
+    pub(super) eval_a: P::ScalarField,
+    pub(super) eval_b: P::ScalarField,
+    pub(super) eval_c: P::ScalarField,
+    pub(super) eval_zw: P::ScalarField,
+    pub(super) eval_s1: P::ScalarField,
+    pub(super) eval_s2: P::ScalarField,
+}
+
+impl<P: Pairing> Round4Proof<P> {
+    fn new(
+        round3_proof: Round3Proof<P>,
+        eval_a: P::ScalarField,
+        eval_b: P::ScalarField,
+        eval_c: P::ScalarField,
+        eval_zw: P::ScalarField,
+        eval_s1: P::ScalarField,
+        eval_s2: P::ScalarField,
+    ) -> Self {
+        Self {
+            commit_a: round3_proof.commit_a,
+            commit_b: round3_proof.commit_b,
+            commit_c: round3_proof.commit_c,
+            commit_z: round3_proof.commit_z,
+            commit_t1: round3_proof.commit_t1,
+            commit_t2: round3_proof.commit_t2,
+            commit_t3: round3_proof.commit_t3,
+            eval_a,
+            eval_b,
+            eval_c,
+            eval_zw,
+            eval_s1,
+            eval_s2,
+        }
+    }
+}
+
+impl<T, P: Pairing> Round4<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: FFTPostProcessing,
+{
+    fn evaluate_poly(
+        driver: &mut T,
+        poly: &FieldShareVec<T, P>,
+        x: &P::ScalarField,
+    ) -> FieldShare<T, P> {
+        let mut res = FieldShare::<T, P>::default();
+        let mut x_pow = P::ScalarField::one();
+        for coeff in poly.clone().into_iter() {
+            let tmp = driver.mul_with_public(&x_pow, &coeff);
+            res = driver.add(&res, &tmp);
+            x_pow *= x;
+        }
+        res
+    }
+    pub(super) fn round4(self) -> PlonkProofResult<Round5<T, P>> {
+        let Self {
+            mut driver,
+            domains,
+            challenges,
+            proof,
+            polys,
+            data,
+        } = self;
+        // STEP 4.1 - Compute evaluation challenge xi \in F_p
+        let mut transcript = Keccak256Transcript::<P>::default();
+        transcript.add_scalar(challenges.alpha);
+        transcript.add_point(proof.commit_t1.into());
+        transcript.add_point(proof.commit_t2.into());
+        transcript.add_point(proof.commit_t3.into());
+        let xi = transcript.get_challenge();
+        let xiw = xi * domains.roots_of_unity[data.zkey.power];
```

Comment:
do we need other RoU other than [0] and [power] otherwise we could just store those?

Reply:
we need [2], [power + 2]

Reply:
so just 4, might be worth doing then? otherwise we keep a vec of len domain size around instead over the whole protocol. Idk if it gets cloned much